File: 0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        ans = [0]
        # multiplying by 2 just adds a 0 => ans[n] = ans[2n]
        # ans[2n+1] = ans[n]+1, the +1 is caused by the odd number
        for x in range(1, n+1):
            ans.append(ans[x>>1]+x%2)
        
        return ans

File: 0338-counting-bits/0338-counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> ans = {0};
        for (int i=1; i<=n; i++)
        {
            ans.push_back(ans[i/2]+i%2);
        }
        
        return ans;
    }
};